File: python-learning-path/practise/fizz_buzz.py
"""
Exercise 10: FizzBuzz

A staple programming challenge to test basic control flow.
"""

def main():
    # Loop from 1 to 50
    for i in range(1, 51):
        
        # Rule: Multiples of both 3 and 5 -> FizzBuzz
        # Check this specific condition FIRST.
        if i % 3 == 0 and i % 5 == 0:
            print("FizzBuzz")
            
        # Rule: Multiples of 3 only -> Fizz
        elif i % 3 == 0:
            print("Fizz")
            
        # Rule: Multiples of 5 only -> Buzz
        elif i % 5 == 0:
            print("Buzz")
            
        # Otherwise, just print the number
        else:
            print(i)

if __name__ == "__main__":
    main()

File: java-learning-path/practise/FizzBuzz.java
/**
 * Exercise 10: FizzBuzz
 * 
 * LOGIC:
 * - Check divisibility by both 3 AND 5 FIRST (the most specific condition).
 * - Then check 5 or 3 individually.
 */
public class FizzBuzz {
    public static void main(String[] args) {
        for (int i = 1; i <= 50; i++) {
            // Condition for multiples of both 3 and 5 (e.g., 15, 30, 45)
            // Tip: (i % 3 == 0 && i % 5 == 0) is the same as (i % 15 == 0)
            if (i % 3 == 0 && i % 5 == 0) {
                System.out.println("FizzBuzz");
            } 
            // Condition for multiples of 3 only
            else if (i % 3 == 0) {
                System.out.println("Fizz");
            } 
            // Condition for multiples of 5 only
            else if (i % 5 == 0) {
                System.out.println("Buzz");
            } 
            // Otherwise, print the number
            else {
                System.out.println(i);
            }
        }
    }
}
